cmake/multi_image.cmake

```diff
@@ -23,6 +23,278 @@ if(IMAGE_NAME)
+  list(FILTER VARIABLES_CACHED INCLUDE REGEX ${regex})
+
+  foreach(var_name
+      ${VARIABLES}
+      ${VARIABLES_CACHED}
+      )
+    # This regex is guaranteed to match due to the filtering done
+    # above, we only re-run the regex to extract the part after
+    # '_'. We run the regex twice because it is believed that
+    # list(FILTER is faster than doing a string(REGEX on each item.
+    string(REGEX MATCH "^${child_image}_(.+)" unused_out_var ${var_name})
+
+    # When we try to pass a list on to the child image, like
+    # -DCONF_FILE=a.conf;b.conf, we will get into trouble because ; is
+    # a special character, so we escape it (mucho) to get the expected
+    # behaviour.
+    string(REPLACE \; \\\\\; val "${${var_name}}")
+
+    list(APPEND child_image_cmake_args
+      -D${CMAKE_MATCH_1}=${val}
+      )
+  endforeach()
+
+  unset(${output_cmake_args} PARENT_SCOPE)
+  set(${output_cmake_args} "${child_image_cmake_args}" PARENT_SCOPE)
+endfunction()
+
+# Adds an S1 variant of an image to the build.
+#
+# Required arguments are:
+# SOURCE_DIR - The source dir of the base image.
+#
+# Optional argumnets are:
+# BASE_IMAGE - The name of the base image which the variant image is based on.
+function(add_s1_variant)
+  set(oneValueArgs BASE_IMAGE SOURCE_DIR)
+  cmake_parse_arguments(ASV "" "${oneValueArgs}" "" ${ARGN})
+
+  if (NOT ASV_SOURCE_DIR)
+    message(FATAL_ERROR "Missing parameter, required: SOURCE_DIR")
+  endif()
+
+  set(s1_override_conf
+    ${ZEPHYR_NRF_MODULE_DIR}/subsys/bootloader/image/is_s1_image.conf)
+
+  add_image_variant(
+    IMAGE_NAME s1_image
+    BASE_IMAGE ${ASV_BASE_IMAGE}
+    EXTRA_CONFIG ${s1_override_conf}
+    SOURCE_DIR ${ASV_SOURCE_DIR}
+    )
+endfunction()
+
+# Adds a variant of an image to the build.
+# The configuration passed to the variant image is a copy of that sent to the
+# base image, except what is passed to the EXTRA_CONFIG argument.
+#
+# Required arguments are:
+# IMAGE_NAME - The name of the variant image
+# SOURCE_DIR - The source dir of the base image.
+# EXTRA_CONFIG - The extra configuration fragment to pass to the variant image.
+#
+# Optional argumnets are:
+# BASE_IMAGE - The name of the base image which the variant image is based on.
+function(add_image_variant)
+
+  set(oneValueArgs BASE_IMAGE IMAGE_NAME SOURCE_DIR EXTRA_CONFIG)
+  cmake_parse_arguments(AIV "" "${oneValueArgs}" "" ${ARGN})
+
+  if (NOT AIV_IMAGE_NAME OR NOT AIV_SOURCE_DIR OR NOT AIV_EXTRA_CONFIG)
+    message(FATAL_ERROR "Missing parameter, required: IMAGE_NAME SOURCE_DIR EXTRA_CONFIG ")
+  endif()
+
+  if (AIV_BASE_IMAGE)
+    # A base image has been specified, this means that we need to copy the
+    # configuration of a child image (as opposed to copying the configuration
+    # from the top level parent image (that is, 'app').
+
+    set(base_image_configuration "${${AIV_BASE_IMAGE}_IMAGE_CMAKE_ARGS}")
+  else()
+    # No base image is specified, this means that the base image is the top
```

Comment:
Can't remember using the term base image, is it the same as parent image?

Reply:
It is introduced in the docs for these functions, it is the image which the variant is based upon.